src/imitation/algorithms/tabular_irl.py

```diff
@@ -130,144 +140,178 @@ def squeeze_r(r_output: th.Tensor) -> th.Tensor:
+
+    def set_demonstrations(self, demonstrations: MCEDemonstrations) -> None:
+        if isinstance(demonstrations, np.ndarray):
+            # Demonstrations are an occupancy measure
+            assert demonstrations.ndim == 1
+            self.demo_state_om = demonstrations
+        else:
+            # Demonstrations are either trajectories or transitions;
+            # we must compute occupancy measure from this.
+            self.demo_state_om = np.zeros((self.env.n_states,))
+
+            if isinstance(demonstrations, Iterable):
+                first_item = next(iter(demonstrations))
+                if isinstance(first_item, types.Trajectory):
+                    # Demonstrations are trajectories.
+                    for traj in demonstrations:
+                        # TODO(adam): vectorize?
+                        cum_discount = 1
+                        for obs in traj.obs:
+                            self.demo_state_om[obs.astype(bool)] += cum_discount
+                            cum_discount *= self.discount
+                else:
+                    # Demonstrations are a Torch DataLoader or other Mapping iterable
+                    for batch in demonstrations:
+                        self._set_demo_from_obs(batch["obs"])
+
+            elif isinstance(demonstrations, types.TransitionsMinimal):
+                self._set_demo_from_obs(demonstrations.obs)
+            else:
+                raise TypeError(
+                    f"Unsupported demonstration type {type(demonstrations)}",
+                )
+
+            self.demo_state_om /= self.demo_state_om.sum()  # normalize
+
+    def train(self, max_iter: int = 1000) -> np.ndarray:
+        """Runs MCE IRL.
 
         Args:
-            obs_dim: size of input observations.
-            hiddens: list of zero or more hidden layer sizes.
-            activation: constructs an activation to insert between hidden
-                layers (e.g. you could supply `nn.ReLU`, `nn.Tanh`, etc.`).
+            max_iter: The maximum number of iterations to train for. May terminate
+                earlier if `self.linf_eps` or `self.grad_l2_eps` thresholds are reached.
+
+        Returns:
+            State occupancy measure for the final reward function. `self.reward_net`
+            and `self.optimizer` will be updated in-place during optimisation.
         """
-        super().__init__()
-        layer_list = []
-        prev_h_size = obs_dim
-        for h_size in hiddens:
-            layer_list.append(nn.Linear(prev_h_size, h_size))
-            layer_list.append(activation())
-            prev_h_size = h_size
-        layer_list.append(nn.Linear(prev_h_size, 1))
-        self.layers = nn.Sequential(*layer_list)
-
-    def forward(self, obs: th.Tensor) -> th.Tensor:
-        return th.squeeze(self.layers(obs), 1)
+        # use the same device and dtype as the rmodel parameters
+        a_param = next(self.reward_net.parameters())
+        device = a_param.device
```

Comment:
Good idea, I've added a `dtype` property. @qxcv @scottemmons can you remember why we have a default for if the module has no parameters? When would it have no parameters -- I guess a constant zero reward network? I'm not sure what a default `dtype` even should be (I went with `float32`).

Reply:
Yeah, I think there is (or was) a constant reward network used for tests somewhere. This was to get around that.
(also generally I think it's a good idea not to assume that `nn.Module`s have any parameters, since there are many built-in ones that are parameter free, like `nn.ReLU`)

Reply:
I think I added that default in the `RewardNet` refactor, at least it's not yet there in this commit: https://github.com/HumanCompatibleAI/imitation/blob/777085d16bf62d63020bc7d0f5d6a740728ff4ad/src/imitation/rewards/reward_nets.py#L185
My only motivation was that it shouldn't raise an error if a `RewardNet` doesn't have parameters for some reason, I didn't have a specific case in mind. But maybe I'm also misremembering and saw it somewhere else (SB3?) and just used it because of that

Reply:
You're right, it was added in https://github.com/HumanCompatibleAI/imitation/pull/311/files#diff-58b170ba20f1518bab467e4e002302c8a873e5e58b9319a4de15231c38766b3cR87 I think it's nice to have a default as a kind of defensive coding, but I'm going to not worry too much what the default is for now until we have a concrete use case. Mildly worried that this is going to cause silent failures down the line, but I can't think of any concrete case (if there are no parameters, shouldn't matter too much what you assume about their device/dtype...)